### DISCUSSION TITLE
Option to replace the core search provider

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8111] [Ideas]
> Created by: Scott Grogan (sgrogan@weareratio.com) on 2015/04/01 21:57:24 +0000
> Votes at time of UserVoice import: 52

-----------------------------------

Currently CraftCMS relies on a core SearchService component backed by the MySQL database. It would be wonderful to be able to replace this service with one backed by ElasticSearch, Lucene, or any other search provider.

**Comment by narration-sd** [comment]
Great idea -- another 'having nice things' ;)

+1, since I don't find a button....??....ah....

**Comment by lindseydiloreto** [comment]
+1 Would love to see an Algolia driver for Craft CP!

**Comment by phoob** [comment]
+100 :)

**Comment by sjelfull** [comment]
Need this!

**Comment by michaelrog** [comment]
_Replacing_ the core search provider is probably not necessary. Craft's search functionality is pretty tightly coupled with the architecture of its internal index. It'd be difficult to ensure/enforce that an external service would obey the same [search syntax](https://craftcms.com/docs/searching#supported-syntaxes), indexing behaviors, and [config options](https://craftcms.com/docs/config-settings#defaultSearchTermOptions) as Craft's native search provider, which could lead to confusion and support issues.

However, in Craft 3, it'll be simple enough [add user-defined behaviors](https://github.com/craftcms/cms/pull/1856) to components (e.g. ElementQuery classes). Such a Behavior could provide an alternate search method/filter that fetches a list of element IDs from an external provider and filters a query by those IDs.

**Comment by yenyrubiano** [comment]
+1 for Swiftype

**Comment by brandonkelly** [comment]
@michaelrog we would most likely still do the query parsing, and hand a [SearchQuery](https://docs.craftcms.com/api/v3/craft-search-searchquery.html) object off to the search driver. So the actual search syntax would remain consistent regardless of the search driver you’re using - it would just be up to the driver to support the SearchQuery as best as it can.

**Comment by phillmorgan28** [comment]
Hi All

Is this functionality still on your radar? I'm looking to replace the core search functionality with an in house Elasticsearch driver, but until support for 3rd party drivers is added I'm pretty dead in the water.

Thanks

**Comment by brandonkelly** [comment]
@phillmorgan28 Definitely still something we’re thinking about... probably for 4.0.

**Comment by gojira32** [comment]
Any update on this @brandonkelly @angrybrad?

Search on the orders screen and other entries is glacial. Our customer service agents are busting my balls on making it quicker.

If we can't make instant search quicker by replacing the core search provider with something of our choice (e.g. Algolia, which we use on the front-end) is there a way we can have more traditional static search as opposed to instant search?

E.g. when a customer service agent wants to search for an order or orders with a specific email (jon@doe.com) all they want is the results for jon@doe.com, not something that is similar and even if they did, they don't want the list of results to dynamically change as they type.

I've raised this before, but something like the Magento 1.x sales orders screen which has static search and where you can search by attribute is a must for ecommerce. It allows you to quickly get a snap shot of orders with certain attributes.

E.g.: https://d.pr/i/PgJwDU

Say you have an issue with orders made between a certain date that are paid via PayPal and in NZD currency. There really is no way (that I'm aware of) that you can easily bring up such a list of orders.

**Comment by brandonkelly** [comment]
@gojira32 We will be addressing it in the coming months ;)

**Comment by gojira32** [comment]
@brandonkelly Addressing this via adding the ability to have static search with query building (like above Magento example) or by allowing replacement of the core search provider?

The former would be much more useful for the orders screen.

**Comment by brandonkelly** [comment]
@gojira32 The former

**Comment by keyurshah** [comment]
to have advanced searching and attribute filtering would be great. 

Often I have to export data out and then edit outside in Excel and then import back in using FeedMe. 

In Commerce's products tab, there's not a way I can find to filter out by Enabled / Disabled products. Or Available To Purchase.

Thanks!

**Comment by angrybrad** [comment]
Promising and should look into: https://typesense.org/ and https://github.com/meilisearch/MeiliSearch

hat-tip @timkelty

**Comment by tekstrand** [comment]
Would love to see this in Craft 4.

**Comment by wmdhosting** [comment]
+1
